src/Propulsion.SqlStreamStore/Checkpoint.fs

```diff
@@ -0,0 +1,133 @@
+﻿module Propulsion.SqlStreamStore.Checkpoint
+
+open FSharp.UMX
+open System // must shadow UMX to use DateTimeOffSet
+
+type CheckpointSeriesId = string<checkpointSeriesId>
+and [<Measure>] checkpointSeriesId
+module CheckpointSeriesId = let ofGroupName (groupName : string) = UMX.tag groupName
+
+let [<Literal>] Category = "Sync"
+let streamName (id : CheckpointSeriesId) = FsCodec.StreamName.create Category %id
+
+// NB - these schemas reflect the actual storage formats and hence need to be versioned with care
+module Events =
+
+    type Checkpoint = { at : DateTimeOffset; nextCheckpointDue : DateTimeOffset; pos : int64 }
+    type Config = { checkpointFreqS : int }
+    type Started = { config : Config; origin : Checkpoint }
+    type Pos = { config : Config; pos : Checkpoint }
+    type Snapshotted = { config : Config; state : Checkpoint }
+
+    type Event =
+        | Started of Started
+        | Checkpointed of Pos
+        | Overrode of Pos
+        // Updated events are not actually written to the store when storing in Cosmos (see `transmute`, below)
+        // While we could remove the `nextCheckpointDue` and `config` values, we won't do that, so people can use AnyKnownEvent
+        //  access modes and/or save just load the most recent event
+        | Updated of Pos
+        | [<System.Runtime.Serialization.DataMember(Name = "state-v1")>]
+            Snapshotted of Snapshotted
+        interface TypeShape.UnionContract.IUnionContract
+    // Avoid binding to a specific serializer as a) nothing else is binding to it in here b) it should serialize with any serializer so we defer
+    // let codec = FsCodec.NewtonsoftJson.Codec.Create<Event>()
+
+module Fold =
+
+    type State = NotStarted | Running of Events.Snapshotted
+
+    let initial : State = NotStarted
+    let private evolve _state = function
+        | Events.Started { config = cfg; origin=originState } -> Running { config = cfg; state = originState }
+        | Events.Updated e | Events.Checkpointed e | Events.Overrode e -> Running { config = e.config; state = e.pos }
+        | Events.Snapshotted runningState -> Running runningState
+    let fold : State -> Events.Event seq -> State = Seq.fold evolve
+
+    let isOrigin _state = true // we can build a state from any of the events and/or an unfold
+
+    let private snapshot state =
+        match state with
+        | NotStarted -> failwith "should never produce a NotStarted state"
+        | Running state -> Events.Snapshotted {config = state.config; state = state.state}
+
+    /// We only want to generate a first class event every N minutes, while efficiently writing contingent on the current etag value
+    /// So, we post-process the events to remove `Updated` events (as opposed to `Checkpointed` ones),
+    /// knowing that the state already has that updated folded into it when we snapshot from it
+    let transmute events state : Events.Event list * Events.Event list =
```

Comment:
as discussed out of band, there are multiple high level choices as to how to manage the checkpoints here
a) the above via lets us write an event per hour in Equinox.Cosmos using the Custom access strategy https://github.com/jet/equinox/blob/master/src/Equinox.Cosmos/Cosmos.fs#L1022. Obviously most folks picking SSS will in general not be interested in taking such a dependency solely to manage checkpoints unless they are already using Cosmos
b) If Equinox.SqlStreamStore had a mechanism to mechanism (and transact against with some etag-like-thing optimistic concurrency control mechanism) unfolds per stream, then one could Just use that (it'd be a killer feature, but it's not a 5 minute job to implement, especially if you're looking to do it in a SQL store neutral way)
c) do something bespoke that maps the processing (and loading of) Updated to some custom table reading.

Right now, unless it's going to be option a or b, its debatable wehtehr this checkpoint impl should even be considered a runner

@kimserey do you have any thoughts on good places and means to use for storing checkpoint positions?